### PR TITLE
[Perf][NPU] Reuse the previous step's decode inputs when the batch has not changed

### PR DESCRIPTION
**Pull request by Junbo2002**
## Purpose

`_prepare_inputs` on the NPU AR runner is written for the general case — chunked
prefill, speculative decode, PCP/DCP, M-RoPE, LoRA, prompt embeds — and rebuilds
every per-step tensor from numpy on every call, pushing ~15 separate H2D copies,
because in general every one of them can change.

On a long single-request decode, none of them do. The MiniCPM-o Talker spends
~112 consecutive steps on one request emitting one codec frame each. Across
those steps the request set, the batch shape and the scheduled-token counts are
all constant; only four things move — the token just sampled, the position, the
sequence length, and the KV slot. Everything else (`query_start_loc`,
`req_indices`, `query_pos`, `num_scheduled_tokens`, `num_accepted_tokens`,
`logits_indices`, `query_lens`) is recomputed to the value it already held.

This PR caches the previous step's derived arrays and applies only the deltas.

## What the removed work costs, measured in place

The evidence here is a timer around the function itself, not an end-to-end
delta: on a shared box a sub-5% end-to-end difference is not resolvable, but
"this call took N µs and ran M times per request" is something a reviewer can
re-derive.

Atlas A3 / 910C, stock `minicpmo_4_5.yaml`, Seed-TTS zh, **32 requests at
concurrency 1**, `_prepare_inputs` wrapped with `perf_counter_ns` and a call
counter dumped per pid. Two rounds; round 2 swaps which die each arm runs on and
runs the arms concurrently.

| | calls / run | round 1 control | round 1 with PR | round 2 control | round 2 with PR |
| --- | ---: | ---: | ---: | ---: | ---: |
| stage 0 (Thinker) | 507 | 562.9 ms | 402.2 ms | 629.0 ms | 369.4 ms |
| stage 1 (Talker) | 3591 | 3410.6 ms | 2183.3 ms | 3547.6 ms | 2213.4 ms |
| **total** | **4098** | **3973.5 ms** | **2585.5 ms** | **4176.6 ms** | **2582.8 ms** |
| per request | | 124.2 ms | **80.8 ms** | 130.5 ms | **80.7 ms** |
| reduction | | | **−34.9%** | | **−38.2%** |

Per call, stage 1: **949.6 / 987.9 µs → 607.9 / 616.4 µs.**

**Call counts are identical in every arm** (507 and 3591), so the two arms did
the same work; only the cost of preparing it moved. The fast path served
**3523 of 3591** stage-1 calls (98.1%) and 439 of 507 stage-0 calls (86.6%).

One thing worth noticing in that table: the treated arm reads **2585.5 and
2582.8 ms** across two rounds on two different dies — 0.1% apart — while the
control reads 3973.5 and 4176.6, which are 5.1% apart. Removing the host work
also removes its sensitivity to what else the box is doing.

### End to end, as corroboration only

| | round 1 | round 2 |
| --- | ---: | ---: |
| mean e2el | 1980.3 → 1928.0 ms (−2.64%) | 1914.1 → 1888.9 ms (−1.32%) |
| mean audio RTF | 0.4715 → 0.4596 (−2.51%) | 0.4575 → 0.4514 (−1.33%) |
| total output tokens | 435 → 435 | 435 → 435 |

Negative in both rounds with the dies swapped, and in the same band as the
in-place saving (43-50 ms against a ~1950 ms request is 2.2-2.6%). I would not
ask anyone to accept the end-to-end numbers on their own — this box drifts more
than that — but they agree with the in-place measurement, and the identical
token counts show the output did not change.

## How the reuse stays safe

The rule is deliberately narrow. The fast path runs **only when the immediately
preceding step also prepared a pure single-token decode with the identical
request set.** Any other step — a prefill, a new request, a reordering, a dummy
run — goes through the generic path, which rewrites the shared buffers, and only
the step *after that* becomes eligible again. A cached array therefore cannot
outlive the buffer contents it was cached against.

Three further properties, each with a test:

- **Every gate returns a reason, not a bool.** A bare bool makes a fast path
  that never engages indistinguishable from one that is broken; with reasons,
  the first block is logged once and says which one it was. There is no
  environment switch to disable this, because there is no state in which it is
  the only thing standing between a correct step and a wrong one — an
  unrecognised config, an unfamiliar block-table layout or an unexpected step
  shape all degrade to today's behaviour.
- **The block table is compared, not assumed.** A decode allocates a block once
  per `block_size` tokens, so most steps re-upload a table the device already
  has; the snapshot is a copy rather than a view, or it would compare the live
  buffer against itself.
- **The cache holds no torch tensors.** A tensor created inside
  `torch.inference_mode()` is an inference tensor, and indexing with one in a
  later step raises "Inference tensors cannot be saved for backward". Everything
  cached is a numpy array, an int or an enum.

`cascade_attn_enabled` is deliberately *not* a gate: cascade prefix lengths are
computed in `execute_model` after `_prepare_inputs` returns and are consumed by
`_build_attention_metadata`, which stays generic — `_prepare_inputs` has no
cascade branch. Gating on it kept the fast path off entirely.

## Test Plan

**vLLM Version:** `0.1.dev1+ge5588e49b` (`0.28.0` line)

**vLLM-Omni Commit:** `d1c8beeb180b4617437c460ba5c8f61a4a561a7e`

1. **Unit tests, CPU only** — `tests/platforms/npu/worker/test_decode_prep_fast.py`,
   17 tests. The gates read plain attributes off the runner, the scheduler
   output and the input batch, so they take duck-typed stand-ins: no
   accelerator, no model, no engine. They cover the step gate (plain decode
   allowed; speculative tokens, multi-token steps, a request still on its
   prompt, an empty batch and three pieces of runner state each blocked with
   their own reason), the block-table snapshot (first call records rather than
   claims unchanged, an identical table is recognised, a newly allocated block
   is noticed, the snapshot is a copy not a view, an unreadable layout keeps
   committing), the signature, and the two invariants above.
2. **In-place probe on real hardware** — the two rounds in the table, on an
   Atlas A3 / 910C.

**On which branch the hardware numbers were taken.** This box's image ships a
vLLM that `main` no longer imports against (`CoreEngineLaunch` is missing from
`vllm.v1.engine.utils`), so `main` cannot serve here. The rounds were run on
`ecd9d99da`. The hook point is unchanged between the two: the `_prepare_inputs`
call site in `npu_ar_model_runner.py` is byte-identical in the surrounding
block, and on both branches `_prepare_inputs` resolves to vllm-ascend's
`NPUModelRunner` with nothing caching it. The diff in this PR applies cleanly to
`d1c8beeb1` and its unit tests run there; I could not run a served benchmark on
`main` myself, and would want a reviewer with a current stack to confirm the
end-to-end column.

## Test Result

`tests/platforms/npu/` on `d1c8beeb1`, same box, before and after:

```
main             : 63 passed, 1 skipped
main + this PR   : 80 passed, 1 skipped
```

+17, no new failures, nothing skipped that was not skipped before. The new file
alone:

```
tests/platforms/npu/worker/test_decode_prep_fast.py    17 passed in 0.06s
```

### CODE REVIEW

**Comment by Junbo2002** [comment]
Self-review, as `vllm-omni-review-bot` asks for before reviewers are assigned.

**Blast radius.** One new module (`decode_prep_fast.py`, 375 lines, imported by exactly one file), one new CPU-only test file, and 27 lines in `npu_ar_model_runner.py`: an import, one attribute in `__init__`, and two method overrides that both call `super()`. Nothing else on the NPU path changes, and nothing outside `vllm_omni/platforms/npu/worker/` is touched. Both overridden methods resolve below us in the MRO (`_prepare_inputs` → vllm-ascend's `NPUModelRunner`, `_dummy_run` → `OmniNPUModelRunner`), so the fallback really is the code that runs today.

**The thing a reviewer should decide deliberately.** This caches derived state across engine steps, which is a class of change that fails *silently* when it fails — a stale `logits_indices` does not raise, it samples the wrong row. So the question is not "is the fast path fast" but "can the cache ever outlive what it describes". My argument that it cannot is structural rather than empirical: the fast path requires that the *immediately preceding* call also took a pure single-token decode with the same request set, so anything that rewrites the shared buffers — a prefill, a new request, a reordering, a dummy run — forces a generic step, and only the step after that is eligible again. I would rather a reviewer attack that sentence than the benchmark.

**The three things I would attack first if this landed in my queue:**

1. *Cascade attention is not gated. Is that a hole?* It was the thing I got wrong first, in the other direction: I gated on `cascade_attn_enabled`, and the fast path then never engaged at all. `_prepare_inputs` has no cascade branch — the prefix lengths are computed in `execute_model` *after* it returns and are consumed by `_build_attention_metadata`, which this PR leaves generic. If that is wrong on a configuration I have not run, it is the gate I would add first.

2. *The block-table comparison is an equality check on a live numpy view.* It is, and the snapshot is taken with `.copy()` for exactly that reason — a view would compare the buffer against itself and report "unchanged" for a table that had moved. There is a test that mutates the row after snapshotting and asserts the change is seen. If the rows cannot be read at all (a layout this does not recognise) it returns False and every step commits, as today.

3. *Why no environment switch?* Because a switch is only useful if the safe state is "off", and here the safe state is already reached without one: every gate returns a reason and falls through to the generic path. Adding an env var would mostly add a configuration in which the tested path is not the one running. I will add one if you would rather have the escape hatch.

**What I could not test, stated plainly.** The hardware numbers were taken on `ecd9d99da`, not on `main`. This box's image ships a vLLM that `main` no longer imports against — `CoreEngineLaunch` is gone from `vllm.v1.engine.utils` — so `main` cannot serve here at all. The diff applies cleanly to `d1c8beeb1` and its unit tests pass there (`tests/platforms/npu/` 63 → 80), and the `_prepare_inputs` call site is unchanged between the two branches, but the served benchmark column is from the older base and I would want someone with a current stack to confirm it.

**Why I trust the in-place number more than the end-to-end one.** Both are in the description, but the claim rests on the first: call counts are identical in every arm (507 and 3591), so the two arms did the same work and only the preparation cost moved. The end-to-end deltas (−2.6% and −1.3%, dies swapped) agree with it and the output token count is identical in all four arms, but this box drifts more than 1% on its own and I would not ask anyone to accept them alone.

File: vllm_omni/platforms/npu/worker/decode_prep_fast.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+        views.append(rows[:num_reqs].copy())
+    cache.block_table_snapshot = views
+
+
+def signature_of(runner: Any) -> tuple:
+    input_batch = runner.input_batch
+    return (input_batch.num_reqs, tuple(input_batch.req_ids))
+
+
+def try_fast_prepare(
```
Comment (lishunyang12) [inline vllm_omni/platforms/npu/worker/decode_prep_fast.py:242]: Add fast-versus-generic tests covering token IDs, positions, masks, and block-boundary updates.

File: vllm_omni/platforms/npu/worker/decode_prep_fast.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+        if all(np.array_equal(p, v) for p, v in zip(previous, views)):
+            return True
+    cache.block_table_snapshot = [v.copy() for v in views]
+    return False
+
+
+def snapshot_block_tables(runner: Any, cache: DecodeInputCache, num_reqs: int) -> None:
+    """Record the rows the generic path just committed, so the next fast step
+    can tell whether anything moved."""
+    tables = getattr(runner.input_batch.block_table, "block_tables", None)
```
Comment (lishunyang12) [inline vllm_omni/platforms/npu/worker/decode_prep_fast.py:223]: Share the block-table row collection; both snapshot paths duplicate the same layout checks.
